Python/chapter03/p03_stack_of_plates/miguelHx.py

```diff
@@ -0,0 +1,285 @@
+            raise IndexError('Stack is Empty')
+        return self.top.data
+
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return self
+
+    def __next__(self) -> T:
+        self.index += 1
+        if self.index == self.size or self.current_node is None:
+            self.index = -1
+            self.current_node = self.top
+            raise StopIteration
+        n: T = self.current_node.data
+        self.current_node = self.current_node.next
+        return n
+
+    def __bool__(self) -> bool:
+        """
+        True is returned when the container is not empty.
+        From https://docs.python.org/3/reference/datamodel.html#object.__bool__ :
+        Called to implement truth value testing and the built-in operation bool();
+        should return False or True. When this method is not defined, len() is called,
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self.size > 0
+    
+    def __len__(self) -> int:
+        return self.size
+
+    def __str__(self):
+        if self.size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+
+class SetofStacks:
+
+    def __init__(self):
+        self.current_stack_idx: int = 0
+        self.set_of_stacks: List[MyStack] = [MyStack()]
+        self.stack_threshold: int = 3
+        self.size: int = 0
+    
+    def push(self, item: T) -> None:
+        # threshold check
+        if len(self.set_of_stacks[self.current_stack_idx]) >= self.stack_threshold:
+            # update current_stack_idx
+            self.current_stack_idx += 1
+            # create new stack
+            self.set_of_stacks.append(MyStack())
+        self.set_of_stacks[self.current_stack_idx].push(item)
+        self.size += 1
+        return
+    
+    def _pop(self) -> T:
+        return self.set_of_stacks[self.current_stack_idx].pop()
+    
+    def _pop_stack(self) -> MyStack:
+        return self.set_of_stacks.pop(self.current_stack_idx)
+
+    def pop(self) -> T:
+        """If the current stack is empty
+        after this pop, then decrement current stack index.
+
+        Returns:
+            T: popped item
+        """
+        if self.current_stack_idx == 0:
+            item: T = self._pop()
+        else:
+            s: MyStack = self._pop_stack()
+            item: T = s.pop()
+            self.current_stack_idx -= 1
```

Comment:
what happens if 5 elements are pushed into the stack and then popped? say,

```python
        sos = SetofStacks()
        sos.push(1)
        sos.push(2)
        sos.push(3)
        sos.push(4)
        sos.push(5)
        self.assertEqual(5, sos.pop())
        self.assertEqual(4, sos.pop())
        self.assertEqual(3, sos.pop())
        self.assertEqual(2, sos.pop())
        self.assertEqual(1, sos.pop())
```

Reply:
Good catch, yes, there is a bug here.  I updated the test cases and implemented a fix here:  c60ef5551631b4d1fd880ff2e066e0726ca1410c